### PR TITLE
Py datasets downloaded from openML 

### LINKED ISSUE #514: Python MNIST examples: data not available out of the box
I'd like to make the MNIST python example runable out of the box. 

it expects manually downloaded file MNIST_data/, 
instead:
- use existing `build/ThirdParty/mnist_data/mnist-src/` which is downloaded for the c++ example 
- or use `sklearn.datasets` 
https://scikit-learn.org/0.19/datasets/mldata.html 

I'd prefer the sklearn method, for it allows an easy way to test on other popular image classification data, 

Some possible additional steps: 
- rename file from mnist to image_classification.py 
- decouple names from the mnist use-case, make more generic & useful 
- try with other datasets: iris, cifar, fashion-mnist, ... 
  - provide optimized params for each of them (?)

### PR DESCRIPTION
**Pull request by breznak**
- [x] fetch data for vision experiments using `sklearn.datasets` from openML
- [x] try fashionMNIST - as example works out of the box

Fixes #514 
Next work can focus on extra encoders for vision: #259

### CODE REVIEW

File: py/htm/examples/mnist.py [hunk 1]
```diff
@@ -15,65 +15,51 @@
...
     return train_labels, train_images, test_labels, test_images
 
+def encode(data, out):
+    """
+    encode the (image) data
+    @param data - raw data
+    @param out  - return SDR with encoded data
+    """
+    out.dense = data >= np.mean(data) # convert greyscale image to binary B/W.
+    #TODO improve. have a look in htm.vision etc. For MNIST this is ok, for fashionMNIST in already loses too much information
```
Comment (breznak) [inline py/htm/examples/mnist.py]: OT: but next step in this domain would be looking at better encoders for vision. The current B/W encoding is likely losing too much information

File: py/htm/examples/mnist.py [hunk 2]
```diff
@@ -94,20 +80,17 @@ def load_images(file_name):
...
 def main(parameters=default_parameters, argv=None, verbose=True):
-    parser = argparse.ArgumentParser()
-    parser.add_argument('--data_dir', type=str,
-        default = os.path.join( os.path.dirname(__file__), '..', '..', '..', 'build', 'ThirdParty', 'mnist_data', 'mnist-src'))
-    args = parser.parse_args(args = argv)
 
     # Load data.
-    train_labels, train_images, test_labels, test_images = load_mnist(args.data_dir)
+    train_labels, train_images, test_labels, test_images = load_ds('mnist_784', 10000, shape=[28,28]) # HTM: ~95.6%
+    #train_labels, train_images, test_labels, test_images = load_ds('Fashion-MNIST', 10000, shape=[28,28]) # HTM baseline: ~83%
```
Comment (breznak) [inline py/htm/examples/mnist.py]: it's now this simple to play with a new image classification dataset! 
Even the results for fashionMNIST are not that bad (83% without touching anything)

File: py/htm/examples/mnist.py [hunk 3]
```diff
@@ -132,17 +115,17 @@ def main(parameters=default_parameters, argv=None, verbose=True):
     # Training Loop
     for i in range(len(train_images)):
         img, lbl = random.choice(training_data)
-        enc.dense = img >= np.mean(img) # Convert greyscale image to binary.
+        encode(img, enc)
         sp.compute( enc, True, columns )
-        sdrc.learn( columns, lbl )
+        sdrc.learn( columns, lbl ) #TODO SDRClassifier could accept string as a label, currently must be int
```
Comment (breznak) [inline py/htm/examples/mnist.py]: @ctrl-z-9000-times what do you think about passing `int(hash(lbl))` to c++ in python classifier? That way the label can be anything (hashable).

File: py/htm/examples/mnist.py (hunk 2 above)
Comment (breznak) [reply py/htm/examples/mnist.py]: ..disadvantage is that one hast to be online, and that the dataset is re-downloaded each time. Can we live with that?

File: py/htm/examples/mnist.py (hunk 3 above)
Comment (ctrl-z-9000-times) [reply py/htm/examples/mnist.py]: That will not work.  The labels for the classifier need to be small integers since the classifier uses an array for them instead of a map.  If you give the SDRC the label "999" then the SDRC will make buckets for all the labels between 0 and 999.

File: py/htm/examples/mnist.py (hunk 1 above)
Comment (ctrl-z-9000-times) [reply py/htm/examples/mnist.py]: Thanks for doing this PR Breznak

On the topic of vision: I wrote a eyeball encoder a while ago, and I haven't had the time to incorporate it into this repo.  Its written in python, using numpy and open-cv.  I will see if I can get some examples of what it does and post them on the encoder topic thread.

Comment (breznak) [reply py/htm/examples/mnist.py]: >  wrote a eyeball encoder a while ago

wow! How does it work in principle? I was thinking about some saccadic/sparse encoder for vision. Please do keep us possted, once you have time. 

Btw, I was looking for what could be salvaged from nupic.vision repo... and nothing much, there's some image encoder with gabon filter...
